### PR TITLE
Remove federated seed controller

### LINKED ISSUE #4078: Drop ability of gardenlet to control multiple seeds
**How to categorize this issue?**

/area usability
/kind enhancement
/priority 3

**What would you like to be added**:
Today, the gardenlet can either control exactly one seed (`.seedConfig`) or multiple seeds by providing a label selector (`.seedSelector`).
We would like to drop the latter entirely.

**Why is this needed**:

There are several concerns with the functionality of controlling multiple seeds:

- Security-related checks cannot be performed properly (#1723) 
- Scalability concerns (a single gardenlet can only control limited number of seeds and shoots)
- Complexity in the code (it's more difficult to establish watches against the seed cluster)

Generally, for above reasons, we don't recommend to use this feature in production but only for development scenarios.

⚠️ One consequence of dropping it would be that the local dev setup would also only be able to control exactly one seed.

### PR DESCRIPTION
**Pull request by xrstf**
**How to categorize this PR?**

/area dev-productivity
/kind technical-debt

**What this PR does / why we need it**:
Now that #4306 has removed the SeedSelector field for Gardenlets, the controller that manages the dynamic set of seeds at runtime (the federated seed controller) is not needed anymore. The set of seeds is now fixed for the entire runtime of the gardenlet (and it's always a single seed).

The two controllers managed by the federatedseed ctrl have been promoted to "1st level controllers" (i.e. moved one directory up) and been adjusted accordingly. Anticipating more controller factoring, no further changes have been made.

**Which issue(s) this PR fixes**:
Fixes #4078

**Special notes for your reviewer**:

**Release note**:

```other other
NONE
```

### CODE REVIEW

File: pkg/gardenlet/controller/networkpolicy/controller.go [hunk 1]
```diff
@@ -58,8 +59,18 @@ type Controller struct {
...
+func NewController(ctx context.Context, clientMap clientmap.ClientMap, logger *logrus.Logger, recorder record.EventRecorder, seedName string) (*Controller, error) {
+	seedClient, err := clientMap.GetClient(ctx, keys.ForSeedWithName(seedName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	// if another controller also needs to work with endpoint resources in the Seed cluster, consider replacing this informer factory with a reusable informer factory.
+	// if this informer factory is not bound to the default namespace, make sure that the endpoint event handlers FilterFunc() also filters for the default namespace.
+	seedDefaultNamespaceKubeInformer := kubeinformers.NewSharedInformerFactoryWithOptions(seedClient.Kubernetes(), 0, kubeinformers.WithNamespace(corev1.NamespaceDefault))
+	endpointsInformer := seedDefaultNamespaceKubeInformer.Core().V1().Endpoints()
```
Comment (rfranzke) [inline pkg/gardenlet/controller/networkpolicy/controller.go]: Can we use the informer from the ctrl-runtime here?

Comment (xrstf) [reply pkg/gardenlet/controller/networkpolicy/controller.go]: I adjusted the code to look like the other informers in 3e72e103a. Is that what you meant? :)

File: pkg/gardenlet/controller/factory.go [hunk 2]
```diff
@@ -156,9 +162,14 @@ func (f *GardenletControllerFactory) Run(ctx context.Context) error {
...
-	federatedSeedController, err := federatedseedcontroller.NewFederatedSeedController(ctx, f.clientMap, f.cfg, f.recorder)
+	networkpolicyController, err := networkpolicycontroller.NewController(ctx, f.clientMap, logger.Logger, f.recorder, f.cfg.SeedConfig.Name)
+	if err != nil {
+		return fmt.Errorf("failed initializing NetworkPolicy controller: %w", err)
+	}
+
+	extensionsController, err := extensionscontroller.NewController(ctx, k8sGardenClient, seedClient, f.cfg.SeedConfig.Name, logger.Logger, f.recorder)
 	if err != nil {
-		return fmt.Errorf("failed initializing federated seed controller: %w", err)
+		return fmt.Errorf("failed initializing NetworkPolicy controller: %w", err)
```
Comment (rfranzke) [inline pkg/gardenlet/controller/factory.go]:
```suggestion
		return fmt.Errorf("failed initializing extensions controller: %w", err)
```

File: pkg/gardenlet/controller/networkpolicy/controller.go [hunk 3]
```diff
@@ -58,8 +60,17 @@ type Controller struct {
...
 }
 
-// NewController instantiates a new controller.
-func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDefaultNamespaceKubeInformer kubeinformers.SharedInformerFactory, seedLogger *logrus.Entry, recorder record.EventRecorder, seedName string) (*Controller, error) {
+// NewController instantiates a new networkpolicy controller.
+func NewController(ctx context.Context, clientMap clientmap.ClientMap, logger *logrus.Logger, recorder record.EventRecorder, seedName string) (*Controller, error) {
+	seedClient, err := clientMap.GetClient(ctx, keys.ForSeedWithName(seedName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get garden client: %w", err)
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/controller/networkpolicy/controller.go]: You can pass the `seedClient` directly in `factory.go`, i.e., no need to pass the entire `clientMap`.

File: pkg/gardenlet/controller/networkpolicy/controller.go [hunk 4]
```diff
@@ -69,25 +80,20 @@ func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDef
...
 
 	controller := &Controller{
 		ctx: ctx,
-		log: seedLogger,
+		log: logger,
 		namespaceReconciler: newNamespaceReconciler(
-			seedLogger,
+			logger,
 			seedClient.Client(),
 			endpointsInformer.Lister(),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/networkpolicy/controller.go]: There is no lister no anymore, right? 👀

File: pkg/gardenlet/controller/networkpolicy/controller.go [hunk 5]
```diff
@@ -99,7 +105,7 @@ func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDef
 		endpointsInformer:       endpointsInformer.Informer(),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/networkpolicy/controller.go]:
```suggestion
		endpointsInformer:       endpointsInformer,
```

File: pkg/gardenlet/controller/networkpolicy/controller.go (hunk 1 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/networkpolicy/controller.go]: Yes, thanks! I think there are still some left-overs after this change, see next review! :)

**Review by rfranzke — COMMENTED** [review]
lgtm, just rebase needed

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**